crates/bin/prove_block/src/rpc_utils.rs

```diff
@@ -111,16 +111,75 @@ fn verify_storage_proof(contract_data: &ContractData, keys: &[Felt]) -> Vec<Felt
+        contract_address!("0x00656ca4889a405ec5222e4b0997e5a043902a98cb1f85a039f76f50c000479d"),
+        // Ekubo:core contract address. Source code is not available but `key_not_in_preimage` error is triggered every time it's called
+        contract_address!("0x5dd3d2f4429af886cd1a3b08289dbcea99a294197e9eb43b0e0325b4b"),
+    ];
+
+    let extra_storage_reads = if special_addresses.iter().any(|address| storage_reads.contains_key(address)) {
+        100 * STORED_BLOCK_HASH_BUFFER
+    } else {
+        STORED_BLOCK_HASH_BUFFER
+    };
+
+    if old_block_number >= Felt252::from(extra_storage_reads) {
+        for i in 1..=extra_storage_reads {
+            keys.entry(contract_address!("0x1"))
+                .or_default()
+                .insert((old_block_number - i).try_into().expect("Felt to StorageKey conversion failed"));
+        }
+    }
+}
+
 pub(crate) async fn get_storage_proofs(
     client: &RpcClient,
     block_number: u64,
     tx_execution_infos: &[TransactionExecutionInfo],
     old_block_number: Felt,
 ) -> Result<HashMap<Felt, PathfinderProof>, ClientError> {
     let accessed_keys_by_address = {
-        let mut keys = get_all_accessed_keys(tx_execution_infos);
+        let (mut keys, storage_reads) = get_all_accessed_keys(tx_execution_infos);
+
         // We need to fetch the storage proof for the block hash contract
-        keys.entry(contract_address!("0x1")).or_default().insert(old_block_number.try_into().unwrap());
+        // Include not only old_block_hash but also retrieve the previous 10 values
+        let block_hash_contract_address = contract_address!("0x1");
+        keys.entry(block_hash_contract_address)
+            .or_default()
+            .insert((old_block_number).try_into().expect("Felt to StorageKey conversion failed"));
+
+        // Include extra keys for contracts that trigger get_block_hash_syscall
+        insert_extra_storage_reads_keys(&storage_reads, old_block_number, &mut keys);
+
+        // Within the Starknet architecture, the address 0x1 is a special address that maps block numbers to their corresponding block hashes. As some contracts might access storage reads using `get_block_hash_syscall`, it is necessary to add some extra keys here.
+        // By leveraging the structure of this special contract, we filter out storage_read_values that are greater than old_block_number and add these extra values to the necessary keys from the 0x1 contract address.
+        // It is worth noting that this approach incurs some overhead due to the retrieval of additional data.
+        // let additional_storage_reads: Vec<Felt> = storage_reads.values().flat_map(|vec| vec.iter()).cloned().collect();
+        let additional_storage_reads: Vec<Felt> = storage_reads
+            .values()
+            .flat_map(|vec| vec.iter().cloned().filter(|x| x <= &Felt252::from(block_number)))
+            .collect::<HashSet<Felt>>()
+            .into_iter()
+            .collect();
```

Comment:
What is the rationale for this?
It seems that you are preventing to call storage values (of contract 0x1) that are also called in another contract tree. 

Why just not get all the storage values that we will ask (we obtain this in `insert_extra_storage_reads_keys`, given if we need it or not) and just extend 
`keys.entry(block_hash_contract_address).or_default().extend(`

It seems that we are adding, subtracting and doing some logic that could be much simpler. 

Does it makes sense?